dataset_eval/artistfilter.py

```diff
@@ -0,0 +1,168 @@
+import os
+import sys
+sys.path.append(os.path.join(os.path.dirname(os.path.realpath(__file__)), ".."))
+
+from sqlalchemy import text
+import logging
+fmt = '%(asctime)s - %(levelname)s - %(message)s'
+logging.basicConfig(level=logging.INFO, format=fmt)
+import random
+import collections
+import json
+
+import db
+import db.dataset
+import db.cache
+import config
+
+#db.init_db_engine(config.SQLALCHEMY_DATABASE_URI)
+
+rtoajson = json.load(open("recordingtoartistmap.json"))
+
+def chunks(l, n):
+    """Yield successive n-sized chunks from l."""
+    for i in xrange(0, len(l), n):
+        yield l[i:i+n]
+
+def print_datadict_summary(datadict):
+    counter = collections.Counter()
+    for r, cls in datadict.items():
+        counter[cls] += 1
+    for cls, count in counter.most_common():
+        print "%s\t\t%s" % (cls, count)
+
+def normalise_datadict(datadict, cut_to):
+    """Take a dictionary of groundtruth and cut all classes to
+    `cut_to` items. If a class has fewer items, discard it.
+    Return newdatadict, removed where `removed` is a dictionary
+    of items in datadict that haven't been added to newdatadict"""
+
+    dataset = collections.defaultdict(list)
+    for r, cls in datadict.items():
+        dataset[cls].append(r)
+    newdataset = {}
+    remaining = {}
+    for cls, items in dataset.items():
+        if len(items) > cut_to:
+            sample = random.sample(items, cut_to)
+            for i in sample:
+                newdataset[i] = cls
+            rest = list(set(items)-set(sample))
+            for i in rest:
+                remaining[i] = cls
+    return newdataset, remaining
+
+def recordings_to_artists(recordings):
+    recordingtoartist = {}
+    numrecordings = len(recordings)
+    for recs in chunks(recordings, 1000):
+        recarts = recordings_to_artists_sub(recs)
+        for r, a in recarts.items():
+            if a:
+                recordingtoartist[r] = a
+    return recordingtoartist
+
+def filter(dataset_id, options):
+    dataset = db.dataset.get(dataset_id)
+    datadict = dataset_to_dict(dataset)
+    if options["filter_type"] == "artist":
+        print ("Filtering by artist")
+        train, test = split_groundtruth(datadict)
+        print (train, test)
+    else:
+        train = datadict
+        test = {}
+    if options["normalize"]:
+        print("Normalising")
+        train, remaining = normalise_datadict(train, 450)
```

Comment:
This is terrible. This value should be based on the smallest size class in the training set. Perhaps there should be a limit somewhere in a configuration file.